helix-core/src/syntax/span.rs

```diff
@@ -0,0 +1,511 @@
+        // or fully consume the span.
+        let mut i = self.index;
+        let mut subslices = 0;
+        loop {
+            match self.spans.get_mut(i) {
+                Some(span) if span.start == self.cursor => {
+                    self.event_queue
+                        .push_back(HighlightStart(Highlight(span.scope)));
+                    i += 1;
+
+                    match subslice {
+                        Some(intersect) => {
+                            // If this span needs to be subsliced, consume the
+                            // left part of the subslice and leave the right.
+                            self.range_ends.push(intersect);
+                            span.start = intersect;
+                            subslices += 1;
+                        }
+                        None => {
+                            // If there is no subslice, consume the span.
+                            self.range_ends.push(span.end);
+                            self.index = i;
+                        }
+                    }
+                }
+                _ => break,
+            }
+        }
+
+        // Ensure range-ends are sorted ascending. Ranges which start at the
+        // same point may be in descending order because of the assumed
+        // sort-order of input ranges.
+        self.range_ends.sort_unstable();
+
+        // When spans are subsliced, the span Vec may need to be re-sorted
+        // because the `range.start` may now be greater than some `range.start`
+        // later in the Vec. This is not a classic "sort": we take several
+        // shortcuts to improve the runtime so that the sort may be done in
+        // time linear to the cardinality of the span Vec. Practically speaking
+        // the runtime is even better since we only scan from `self.index` to
+        // the first element of the Vec with a `range.start` after this range.
+        if let Some(intersect) = subslice {
+            let mut after = None;
+
+            // Find the index of the largest span smaller than the intersect point.
+            // `i` starts on the index after the last subsliced span.
+            loop {
+                match self.spans.get(i) {
+                    Some(span) if span.start < intersect => {
+                        after = Some(i);
+                        i += 1;
+                    }
+                    _ => break,
+                }
+            }
+
+            // Rotate the subsliced spans so that they come after the spans that
+            // have smaller `range.start`s.
+            if let Some(after) = after {
+                self.spans[self.index..=after].rotate_left(subslices);
+            }
+        }
+
+        self.event_queue.pop_front()
+    }
+}
+
+struct FlatSpanIter<I> {
+    iter: I,
+}
+
+/// Converts a Vec of spans into an [Iterator] over [HighlightEvent]s
+///
+/// This implementation does not resolve overlapping spans. Zero-width spans are
```

Comment:
Zero-width spans were actually not eliminated but was trivial to implement. (See 36d44f54b330d81ac26cd9dda9b1713a0520d6ae)